virttest/env_process.py

```diff
@@ -1166,9 +1166,29 @@ def preprocess(test, params, env):
 
     if (params.get("auto_cpu_model") == "yes" and
             vm_type == "qemu"):
-        if not env.get("cpu_model"):
-            env["cpu_model"] = cpu.get_qemu_best_cpu_model(params)
-        params["cpu_model"] = env.get("cpu_model")
+        policy_map = {"libvirt_host_model": cpu.get_cpu_info_from_virsh,
+                      "virttest": cpu.get_qemu_best_cpu_info}
+        auto_cpu_policy = params.get("auto_cpu_policy", "virttest").split()
+        for policy in auto_cpu_policy:
+            try:
+                cpu_info = policy_map[policy](params)
+                if cpu_info:
+                    break
+            except Exception as err:
+                logging.error("Failed to get cpu info with policy %s, the "
+                              "error message is \"%s\"" % (policy, err))
+                continue
+        else:
+            test.cancel("Failed to get cpu info with policy %s"
+                        % auto_cpu_policy)
+
+        env["cpu_model"] = cpu_info["model"]
+        env["cpu_flags"] = cpu_info["features"]
+        params["cpu_model"] = env["cpu_model"]
+        if env["cpu_flags"]:
+            cpu_flags = params.get("cpu_model_flags")
+            params["cpu_model_flags"] = cpu.combine_flags(env["cpu_flags"],
+                                                          cpu_flags)
```

Comment:
`env['cpu_flags']`  is a new env params,  what exactly does it mean ?
I wonder do we need to update `env['cpu_flags']`  after combine ?  Indicate the cpu flags that be used in vm finally.

Reply:
Or, no need to save cpu_flags with env, the new design use the cache file strategy, and keep `env['cpu_model']` just to compatible with the older code.